Introduce new PSR-18 based client (#257)
kbsali/php-redmine-api: docs/usage.md

# Usage

> :warning: This file shows example commands for reading, creating, updating and deleting redmine components. As this are only examples, please **make sure, that you do not execute them at your production Redmine instance and may destroy real life content!**

Reviewer: ```suggestion
:warning: This file shows example commands for reading, creating, updating and deleting Redmine data. As these are only examples, please **make sure, that you do not execute them against your production instance as it may break real content!**
```

## Installation

### With Composer

Please make sure you have install `php-remine-api` with an PSR-4 autoloader like [Composer](https://getcomposer.org/).

```bash

> php composer.phar require kbsali/php-redmine-api

Reviewer: ```suggestion
composer require kbsali/php-redmine-api
```

```

This will create a new entry in your `composer.json`.

```diff
{
    "require": {
+        "kbsali/php-redmine-api": "^1.6"
    }
}

```

### ... or without Composer

As an alternative you can download the code with all required dependencies from [php-download.com](https://php-download.com/package/kbsali/redmine-api) and unzip all files in your project folder.

### Start your project

Create your project e.g. in the `index.php` by require the `vendor/autoload.php`

```diff
+<?php
+
+require_once 'vendor/autoload.php';
```

### Instantiate a Redmine Client

You can choose between the navite curl client without other PHP dependencies or the PSR-18 compatible client.

#### Native curl Client `Redmine\Client`

Every Client requires a URL to your Redmine instance and either an Apikey...

```diff
<?php

require_once 'vendor/autoload.php';
+
+// Instantiate with ApiKey
+$client = new Redmine\Client('http://localhost', '1234567890abcdfgh');
```

... or valid username/password.

```diff
<?php

require_once 'vendor/autoload.php';
+
+// Instantiate with ApiKey with Username/Password
+$client = new Redmine\Client('http://redmine.example.com', 'username', 'password');
```

> > :bulb: For security reason it is recommended that you use a ApiKey instead of username/password.

Reviewer: ```suggestion
> :bulb: For security reason it is recommended that you use an ApiKey rather than your username/password.
```

Author: Thank you very much. :heart:  I'm not a native speaker.

```diff
<?php

require_once 'vendor/autoload.php';

// Instantiate with ApiKey
$client = new Redmine\Client('https://redmine.example.com', '1234567890abcdfgh');
+
+// [OPTIONAL] if you want to check the servers' SSL certificate on Curl call
+$client->setCheckSslCertificate(true);
+
+// [OPTIONAL] set the port (it will try to guess it from the url)
+$client->setPort(8080);
+
+// [OPTIONAL] set a custom host
+$client->setCustomHost('https://localhost:8080');

```
#### Psr-18 compatible Client `Redmine\Client\Psr18Client`

The `Psr18Client` requires

- a `Psr\Http\Client\ClientInterface` implementation (like guzzlehttp/guzzle), [see](https://packagist.org/providers/psr/http-client-implementation)
- a `Psr\Http\Message\ServerRequestFactoryInterface` implementation (like nyholm/psr7), [see](https://packagist.org/providers/psr/http-factory-implementation)
- a `Psr\Http\Message\StreamFactoryInterface` implementation (like nyholm/psr7), [see](https://packagist.org/providers/psr/http-message-implementation)
- a URL to your Redmine instance
- an Apikey or username
- and optional a password if you want tu use username/password.

> > :bulb: For security reason it is recommended that you use a ApiKey instead of username/password.

Reviewer: ```suggestion
> :bulb: For security reason it is recommended that you use an ApiKey rather than your username/password.
```

```diff
<?php

require_once 'vendor/autoload.php';
+
+$guzzle = \GuzzleHttp\Client();
+$psr17Factory = new \Nyholm\Psr7\Factory\Psr17Factory();
+
+// Instantiate with ApiKey
+$client = new Redmine\Client\Prs18Client($guzzle, $psr17Factory, $psr17Factory, 'https://redmine.example.com', '1234567890abcdfgh');
+// ...or Instantiate with ApiKey with Username/Password (not recommended)
+$client = new Redmine\Client\Prs18Client($guzzle, $psr17Factory, $psr17Factory, 'https://redmine.example.com', 'username', 'password');
```

##### Guzzle configuration

Because the `Psr18Client` is agnostic about the HTTP client implementation every configuration specific to the transport has to be set to the `Psr\Http\Client\ClientInterface` implementation.

> This means if you want to set any curl settings to guzzle you have multiple ways to set them:

Reviewer: ```suggestion
This means that if you want to set any `cURL` settings to `Guzzle` you have multiple ways to set them:
```
